File: source/weha/wikianalysis/lexer/WikiToken.java
package mo.umac.wikianalysis.lexer;

public class WikiToken extends Token {

	private static final long serialVersionUID = 7724510731264682048L;

	public String displayString;
	
	/**
	 * No-argument constructor
	 */
	public WikiToken() {
		super();
		this.displayString = new String();
	}

	/**
	 * Constructs a new token for the specified Lind.
	 * @param kind
	 */
	public WikiToken(int kind) {
		super(kind);
		this.displayString = new String();
	}

	/**
	 * Constructs a new token for the specified Image and Kind.
	 * @param kind
	 * @param image
	 */
	public WikiToken(int kind, String image) {
		super(kind, image);
		this.displayString = new String(image);
	}

	public WikiToken(int kind, String image, String displayString) {
		super(kind, image);
		this.displayString = new String(displayString);
	}
	
	public WikiToken(Token token) {
		super(token.kind, token.image);
		this.displayString = new String(token.image);
	}
	
	public WikiToken(Token token, String displayString) {
		super(token.kind, token.image);
		this.displayString = new String(displayString);
	}
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((image == null) ? 0 : image.hashCode());
		result = prime * result + kind;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Token other = (Token) obj;
		if (image == null) {
			if (other.image != null)
				return false;
		} else if (!image.equals(other.image))
			return false;
		if (kind != other.kind)
			return false;
		return true;
	}

	@Override
	public String toString() {
		return this.displayString;
	}
	
}

File: weha/source/weha/WikiToken.php
<?php
class WikiToken {

	public $kind;
	public $image;
	public $displayString;
	public $userName;
	public $editionId;
	private $tokenHash;
	
	/**
	 * Constructs a new token for the specified Image and Kind.
	 * @param kind
	 * @param image
	 * @param $displayString
	 */
	public function __construct($kind, $image, $displayString = null, $userName = null, $revisionId = null) {
		$this->kind = $kind;
		$this->image = $image;
		$this->userName = $userName;
		$this->editionId = $revisionId;
		
		if ($displayString === null)
			$this->displayString = $image;
		else
			$this->displayString = $displayString;
		
		$this->tokenHash = uniqid();
	}
	
	public function createWikiToken($token, $displayString = null) {
		$this->kind = $token->kind;
		$this->image = $token->image;
		
		if ($displayString === null)
			$this->displayString = $token->image;
		else
			$this->displayString = $displayString;
	}
	
	public function __toString() {
		return $this->displayString;
	}
	
	public function hashCode() {
		$prime = 31;
		$result = 1;
		$result = $prime * $result + (($this->image == null) ? 0 : crc32($this->image));
		$result = $prime * $result + $this->kind;
		return $result;
	}

	public function equals($obj) {
		if ($this === $obj)
			return true;
		if ($obj === null)
			return false;
		if (get_class() != get_class($obj))
			return false;
		$other = $obj;
		if ($this->image === null) {
			if ($other->image !== null)
				return false;
		} else if ($this->image != $other->image)
			return false;
		if ($this->kind != $other->kind)
			return false;
		return true;
	}
	
	public function getHash(){
		return $this->tokenHash;
	}
	
	
}
